caffeine@patapon.info/mprisMediaPlayer2.js

```diff
@@ -0,0 +1,342 @@
+            dbusName,
+            '/org/mpris/MediaPlayer2',
+            (_player) => this._onPlayerChange()
+        );
+
+        const handlerId = dbusPlayerProxy.connect(
+            'g-properties-changed',
+            (_player) => this._onPlayerChange()
+        );
+
+        this._activePlayers.set(dbusName, {
+            handlerId,
+            playerProxy: dbusPlayerProxy
+        });
+    }
+
+    /**
+     * @param {string} dbusName Name in dbus
+     */
+    _removePlayer(dbusName) {
+        const player = this._activePlayers.get(dbusName);
+        if (!player) {
+            return;
+        }
+        player.playerProxy.disconnect(player.handlerId);
+        this._activePlayers.delete(dbusName);
+    }
+
+    _onPlayerChange() {
+        let isPlaying = false;
+        for (const player of this._activePlayers.values()) {
+            if (player.playerProxy.PlaybackStatus === 'Playing') {
+                isPlaying = true;
+            }
+        }
+        this._isPlaying = isPlaying;
+        this._emitPlayStatus();
+    }
+
+    _onNameOwnerChanged(_proxy, _sender, [name, oldOwner, newOwner]) {
+        if (!name.startsWith(this._mprisPrefix)) {
+            return;
+        }
+        if (oldOwner && !newOwner) {
+            this._removePlayer(name);
+        }
+        if (newOwner && !oldOwner) {
+            this._addPlayer(name);
+        }
+        this._onPlayerChange();
+    }
+
+    /**
+     * Get the dbus name list for mpris players
+     * @returns {string[]}
+     */
+    _getMPlayerApps() {
+        const [names] = this._dbusProxy.ListNamesSync();
+        const mprisPlayers = names.filter((dbusName) =>
+            dbusName.startsWith(this._mprisPrefix)
+        );
+
+        return mprisPlayers;
+    }
+
+    _onDestroy() {
+        this._dbusProxy.disconnectSignal(this._dbusHandlerId);
+        for (const dbusName of this._activePlayers.keys()) {
+            this._removePlayer(dbusName);
+        }
+        this._activePlayers.clear();
+
+        for (const connectId of this._connections.values()) {
+            this.disconnectIsPlaying(connectId);
+        }
+        this._connections.clear();
+    }
+
+    /**
+     * @param {any} params No params needed
+     */
+    constructor(params) {
+        super();
+        // Declare some GObject.Object properties (non typescript has some difficulty with this weird import)
+        if (!this.emit) {
+            this.emit = super.emit;
+        }
+        if (!this.connect) {
+            this.connect = super.connect;
+        }
+        if (!this.disconnect) {
+            this.disconnect = super.disconnect;
+        }
+
+        if (!(params instanceof PrivateContructorParams)) {
+            throw new TypeError(
+                'MprisMediaPlayer2 is not constructable. ' +
+                'Use `MprisMediaPlayer2.Get()` and `MprisMediaPlayer2.Destroy()`'
+            );
+        }
+
+        this._DBusProxy = Gio.DBusProxy.makeProxyWrapper(DBusInterface);
```

Comment:
Can this be local to this function? It doesn't seem to be referenced outside of it

Reply:
It could be. It's a dependency of this class that methods could use if needed... This is just standard initialization of the class.

Reply:
It also add the `@typedef` on the property.

https://github.com/PVermeer/gnome-shell-extension-caffeine/blob/afec6f609f215aa82cfa325eee899be8f917a829/caffeine%40patapon.info/mprisMediaPlayer2.js#L20-L29

I recommend to keep it.

Reply:
I removed it anyways.